Feature: Filter on process list

stijnveenman opened the pull request:
Add the ability to filter on process list using `/` (vim keybinding). 
Can be confirmed with enter (has the same issue as https://github.com/F1bonacc1/process-compose/pull/135)
Or cancelled with Escape. On start it resets the search, and searches as you type to see the result. 
If cancelled early it will reset the search. 


https://github.com/F1bonacc1/process-compose/assets/126677968/83be52e7-7a94-4bd8-ab8d-5866c2f566f7

I did run into one issue where, if the search results in an empty list, it hangs the process. The processes are still running in the background (seen by the logs updating). But the TUI doesn't respond anymore, i haven't been able to figure out what the reasoning is behind this. If i click escape quickly (to cancel the search), the problem doesn't show up.

F1bonacc1 commented on src/tui/proc-filter.go:
```diff
@@ -0,0 +1,68 @@
+		pv.searchProcess(searchTerm, isRegex, caseSensitive)
+	}
+	f.GetFormItem(0).(*tview.InputField).SetChangedFunc(func(_ string) {
+		searchFunc()
```
Can you pass the text to the `searchFunc` here?
```golang
searchFunc := func(searchTerm string) {
  caseSensitive := f.GetFormItem(1).(*tview.Checkbox).IsChecked()
  isRegex := f.GetFormItem(2).(*tview.Checkbox).IsChecked()
  
  pv.searchProcess(searchTerm, isRegex, caseSensitive)
}
f.GetFormItem(0).(*tview.InputField).SetChangedFunc(func(text string) {
  searchFunc(text)
})

F1bonacc1 commented on src/tui/proc-filter.go:
```diff
@@ -0,0 +1,68 @@
+	f.SetButtonsAlign(tview.AlignCenter)
+	f.SetTitle("Search Process")
+
+	f.AddInputField("Search For", pv.logsText.getSearchTerm(), fieldWidth, nil, nil)
```
```golang
pv.logsText.getSearchTerm()
```
Looks like a copy-paste from the log search 😉
It should be the current search term if wasn't reset.

F1bonacc1 commented on src/tui/proc-table.go:
```diff
@@ -46,6 +47,12 @@ func (pv *pcView) fillTableData() {
 			continue
 		}
+
+		if pv.processRegex != nil && !pv.processRegex.MatchString(state.Name) {
```
I think `pv.processRegex` should be protected with a mutex.
It's being read and written from 2 different threads.

F1bonacc1 reviewed (changes requested):
This looks great. Users with many processes will ask themself how did they live without this feature!

A few small changes and we are good to go.

stijnveenman commented:
I am one of said users with many processes, which is why i wanted this feature 😄 

I've fixed the remarks. And added a mutex. But it doesn't seem to fix the issue. The issue also doesn't happen during the search (when the regex is being written). But happens only if you submit the search with an empty result set. Which makes me suspect it's got something to do with the log window updating and doing something with the selected process, when no process is selected

F1bonacc1 commented on src/tui/proc-table.go:
```diff
@@ -209,6 +221,51 @@ func (pv *pcView) setTableSorter(sortBy ColumnID) {
+}
+
+func (pv *pcView) searchProcess(search string, isRegex, caseSensitive bool) error {
+	pv.processRegexMtx.Lock()
```
That's a good start that you added a mutex to the regex assignment, but the use of the `pv.processRegex` is still unprotected during match.
What you'll probably want to do is implement 2 funcs:

```golang
func (pv *pcView) setProcRegex(reg *regexp.Regexp) {
	pv.processRegexMtx.Lock()
	defer pv.processRegexMtx.Unlock()
	//it can be nil when you want to reset
	pv.processRegex = reg
}

func (pv *pcView) matchProcRegex(procName string) bool {
	pv.processRegexMtx.Lock()
	defer pv.processRegexMtx.Unlock()
	if pv.processRegex != nil {
		return pv.processRegex.MatchString(procName)
	}
	return false
}
```

stijnveenman replied:
That makes sense, i do wonder. Does this need a mutex? 
The `proc-filter` form creates a new regex for every input, and saves the pointer to `pv.processRegex`.
When the `fillTableData` gets the current pointer to the current regex. If another search input it done during the `fillTableData` call, a new regex is created and written to `pv`. `fillTableData` still points to the old regex though, and this regex is not garbage collected. So at worst its slightly outdated?

F1bonacc1 replied:
2 different threads writing to the same memory location, even for a new value assignment, is a recipe for disaster :)

F1bonacc1 commented on src/tui/proc-table.go:
```diff
@@ -46,13 +47,24 @@ func (pv *pcView) fillTableData() {
+		pv.procTable.SetSelectable(false, false)
+	} else {
+		pv.procTable.SetSelectable(true, false)
+	}
```
I think you encountered my bug, which I already fixed in my env.
Instead of this block, can you try:
```golang
if pv.procTable.GetRowCount()-1 > row - 1{
```
Notice the `row - 1`
In the line below.

stijnveenman replied:
I added this block as a workaround for the issue in `tview`. It seems to break if the currently selected row is removed from the table. Which i was also able to replicate in isolation. 

At one point i tried to fix it by calling `table.clear()` before rendering the rows, which means it didn't have to clean up the remaining rows. And it still had the same issue (with the lines you mentioned removed). 

I tried it with this change and run into the same issue.  (to test it)
- enter a filter that results in no processes
- press j or k

F1bonacc1 replied:
That's strange. I couldn't replicate it in my environment by, for example, hiding a selected namespace `Ctrl+G`.
Even if it's the last row, leaving me with 0 rows (except the title row).

stijnveenman replied:
Interesting. How did you filter all rows using namespaces? Doesn't every namespace in the list have at least one project?

stijnveenman replied:
Weird. If i
- use the process filter to filter everything but nginx. 
- filter away nginx using namespace filter
- press j
i run into the same issue (notice it only crashes when i actually click j)

F1bonacc1 replied:
You are right...
Once I click `j` the same happens to me.

F1bonacc1 replied:
It doesn't crash though, for me the process compose is stuck and consumes a lot of CPU.

stijnveenman replied:
Yes sorry, that's what i ment by "crashing" , actually the background processes also keep running fine. (As seen by the logs updating by looking at the logfile directly)

F1bonacc1 replied:
I attached a goroutine profiler. It looks like the table is stuck in this loop:
https://github.com/rivo/tview/blob/861aa94d61c899b32a1304d61b840d0178ece408/table.go#L1344

```
#	0x65ae14	github.com/rivo/tview.(*Table).InputHandler.func1.2+0x54			/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/table.go:1375
#	0x65aca3	github.com/rivo/tview.(*Table).InputHandler.func1.5+0xa3			/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/table.go:1440
#	0x65a5a4	github.com/rivo/tview.(*Table).InputHandler.func1+0x4e4				/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/table.go:1598
#	0x65a08f	github.com/rivo/tview.(*Table).InputHandler.(*Box).WrapInputHandler.func2+0x4f	/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/box.go:167
#	0x63bb16	github.com/rivo/tview.(*Grid).InputHandler.func1+0x1b6				/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/grid.go:277
#	0x63b92f	github.com/rivo/tview.(*Grid).InputHandler.(*Box).WrapInputHandler.func2+0x4f	/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/box.go:167
#	0x650801	github.com/rivo/tview.(*Pages).InputHandler.func1+0xa1				/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/pages.go:311
#	0x65072f	github.com/rivo/tview.(*Pages).InputHandler.(*Box).WrapInputHandler.func2+0x4f	/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/box.go:167
#	0x627737	github.com/rivo/tview.(*Application).Run+0x557					/home/eugene/go/pkg/mod/github.com/rivo/tview@v0.0.0-20240101144852-b3bd1aa5e9f2/application.go:344
#	0xb3a3fd	github.com/f1bonacc1/process-compose/src/tui.SetupTui+0x13d			/home/eugene/projects/go/process-compose/src/tui/view.go:474
#	0xb7ebb8	github.com/f1bonacc1/process-compose/src/cmd.startTui+0x178			/home/eugene/projects/go/process-compose/src/cmd/project_runner.go:92
```

stijnveenman replied:
Given this remaining issue, is there anything else in this PR that still needs to be addressed? 
Do you think this workaround should be used for now? GIven that #133 will have a similar issue? Or wait for an upstream fix?

F1bonacc1 replied:
I already merged your PR locally.
I believe I will release a new version over the weekend.
Does this timeline work for you?

stijnveenman replied:
Awesome! no rush, i was just wondering if there were any more changes needed 👌
